### PR TITLE
Fix joinpath and splitpath for DXPaths

### PR DESCRIPTION
**Pull request by anujkumar93**
@pkaleta @jtratner 

This PR fixes bugs in joinpath and splitpath for DXPath.

## Background
Stor had a bug for DXPaths: `stor.join('dx://spar_dts_regression_2019_01', 'test')` resulted in `ValueError: Project is required to construct a DXPath`. This was because the joinpath and splitpath defaulted to posixpath behavior for all OBS. This needed to be fixed, since all the parts of the sequence above in the join argument are valid.

## Changes
- Fixed the joinpath and splitpath function in DXPath
- Added tests for both.

## Tested via:
- Test suite was run successfully

## TODO:
- Run integration tests once all bugs are fixed / features are added

### CODE REVIEW

**Comment by jtratner** [comment]
let's close #104 then. Also needs doc note about `splitpath`

**Comment by anujkumar93** [comment]
@jtratner added to release notes. I don't have write permission to counsyl/stor so you'll probably have to merge https://github.com/counsyl/stor/pull/104 .

**Comment by pkaleta** [comment]
@anujkumar93 I've merged https://github.com/counsyl/stor/pull/104 and merged master back here.

**Comment by pkaleta** [comment]
Nice tests @anujkumar93! This LGTM.

File: stor/tests/test_dx_path_compat.py [hunk 1]
```diff
@@ -143,3 +144,77 @@ def test_properties(self):
...
+            'dx://project-123456789012345678901234:file-123456789012345678901234')
+
+    def test_splitpath(self):
+        f = DXPath('dx://project:/prefix/file')
+        assert f.splitpath() == (DXPath("dx://project:/prefix"), 'file')
+        f = DXPath('dx://project:/prefix')
+        assert f.splitpath() == (DXPath("dx://project:"), 'prefix')
+        f = DXPath('dx://project:/')
+        assert f.splitpath() == (DXPath("dx://project:"), '')
+        f = DXPath('dx://project:')
```
Comment (pkaleta) [inline stor/tests/test_dx_path_compat.py]: Is this the correct behavior? IMHO this should return the same as the previous line.

```python
In [31]: f=stor.Path('dx://project:')

In [32]: f.splitpath()
Out[32]: (PosixPath('dx:'), u'project:')
```

File: stor/dx.py [hunk 2]
```diff
@@ -1222,3 +1228,13 @@ def canonical_path(self):
 
     def normpath(self):
         return self.path_class(self.drive + self.project + ':' + (self.resource or ''))
+
+    def splitpath(self):
+        """Wrapper around base splitpath function which calls splitpath on the normpath of self
+        """
+        path_w_slash = self
```
Comment (pkaleta) [inline stor/dx.py]: Why is this called `path_w_slash`. Does this always have slash at the end?

File: stor/tests/test_dx_path_compat.py [hunk 3]
```diff
@@ -143,3 +144,77 @@ def test_properties(self):
...
+        assert f.splitpath() == (stor.Path("dx:"), 'project:')
+        f = DXPath('dx://project')
```
Comment (pkaleta) [inline stor/tests/test_dx_path_compat.py]: ditto

File: stor/tests/test_dx_path_compat.py [hunk 4]
```diff
@@ -143,3 +144,78 @@ def test_properties(self):
...
+        assert f.splitpath() == (DXPath("dx://project-123456789012345678901234:"), 'prefix')
+        # This won't resolve to a canonical project:
+        f = DXPath('dx://project-123456789012345678901234:prefix')
+        assert f.splitpath() == (DXPath("dx://project-123456789012345678901234:"), 'prefix')
+        f = DXPath('dx://project-123456789012345678901234:/')
+        assert f.splitpath() == (DXCanonicalPath("dx://project-123456789012345678901234:"), '')
+        f = DXPath('dx://project-123456789012345678901234:')
+        assert f.splitpath() == (stor.Path('dx:'), 'project-123456789012345678901234:')
+        f = DXPath('dx://project-123456789012345678901234')
+        assert f.splitpath() == (stor.Path('dx:'), 'project-123456789012345678901234')
```
[the remark this answers is not in the document]
Comment (anujkumar93) [reply stor/tests/test_dx_path_compat.py]: Hmm, so splitpath() on DXPaths will always have a DXPath as a parent. As opposed to Swift and S3 paths like `swift://tenant` which errors. That's cleaner, fixed.

**Comment by pkaleta** [comment]
This LGTM % failing test, can you PTAL?

**Comment by anujkumar93** [comment]
Yeah, it was a random Travis error that went away after I retriggered. Please feel free to merge this.

**Comment by pkaleta** [comment]
Thanks @anujkumar93!

**Pull request merged**